Docs/fixed-folders

NearW opened the pull request:
# Add documentation on how to generate custom cc.jsons with the fixed attribute 

_**Please read the [CONTRIBUTING.md](https://github.com/MaibornWolff/codecharta/blob/main/CONTRIBUTING.md) before opening a PR.**_

## Description

This is the first PR adding the docs. Adding the docs first improves our understanding of the overall feature.

- Adds an image to visualize margin and the folders
- Adds documentation on how to generate **valid** custom `cc.jsons`

## Sceenshots

![image](https://user-images.githubusercontent.com/12533626/90397992-fe0ae880-e098-11ea-80e4-8fcfd5d8cb86.png)

NearW commented:
Based on my implementation, the margin might not be required in the `.cc.json`.

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
```diff
@@ -0,0 +1,98 @@
+The `fixed`-attribute includes a coordinate `x` and `y`. These represent the top-left corner of the folder. `width`
+defines the length in x-direction and `height` defines the length in y-direction.
+
+Note, that these numbers represent percentages. The overall map size is therefore `100`.
```
```suggestion
The property values must be integers in the range between 0 and 100. They represent the size of the value in percent.
```

NearW replied:
```suggestion
The property values must be numbers in the range between 0 and 100. They represent the size of the value in percent.
```

You can also specify `x: 1.5`

BridgeAR replied:
Ha, alright.

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
````diff
@@ -0,0 +1,98 @@
+ }
+```
+
+The `fixed`-attribute includes a coordinate `x` and `y`. These represent the top-left corner of the folder. `width`
+defines the length in x-direction and `height` defines the length in y-direction.
````
We might not need this if we rename the properties to something more intuitive?

NearW replied:
True. Should we keep `width` and `height` or should we use `width` and `length`?

BridgeAR replied:
We should keep `height`.

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
````diff
@@ -0,0 +1,98 @@
+}
+```
+
+Notice, that `folder_1` start at `x:5`, `y:5`, because we apply a margin of `5` around the map.
````
Is the margin always present and is it fixed at the value 5?

NearW replied:
No. This is really confusing here. I will rephrase this again.

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
```diff
@@ -0,0 +1,98 @@
+In order to build a valid custom `.cc.json`, these rules must be followed:
+
+-   The value `x + width` or `y + height` must be in a range of `[0, 100]`.
+-   Folders can't overlap. Applying a margin between folders is recommended.
```
The margin can not be changed by the user, right? If that's the case, the second sentence seems to be redundant?

NearW replied:
The margin is determined by the space that is left over after you defined your `fixed`-object. The margin can be different on every side.

Maybe `Leaving some space between folders for visibility reasons is recomended.` is better?

BridgeAR replied:
Smart. Even though we can't determine all bad input cases due to that. Seems like it's a well made consideration.

BridgeAR replied:
Your suggestion sound pretty good. I'd go with

```suggestion
-   Folders may not overlap.
-   Leaving space between folders (for visibility reasons) is recomended.
```

NearW commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
```diff
@@ -0,0 +1,98 @@
+
+# Fixate folders
+
+In order to prevent folders from changing locations, we introduced a new attribute to the `.cc.json` called `fixed`.
```
I just talked about the same thing with @Richargh 
We should either go with `fixedPosition` or `fixedLocation`. Position is more suiting imo.

NearW commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
````diff
@@ -0,0 +1,98 @@
+
+```json
+ "fixed": {
+    "x": 60,
+    "y:": 40,
````
`left` and `top` sounds great.

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
```diff
@@ -0,0 +1,101 @@
+The [Squarified-Tree-Map-Algorithm](https://www.win.tue.nl/~vanwijk/stm.pdf) generates a layout for the visualized map.
+The layout varies depending on the area of each node. Adding a new file to a folder might increase the area of the parent folder,
+so that the algorithm decides that the parent-folder needs to be relocated. Another way to follow this problem is by changing the area-metric.
+Notice, that most folders will not be located at the same position as before.
```
```suggestion
Thus, folders may not be located at the same position anymore as before.
```

Ideally move the sentence before the `Another way to follow...`

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
```diff
@@ -0,0 +1,101 @@
+
+![Example]({{site.baseurl}}/assets/images/posts/how-to/fixate-folders/fixate-folder-example.jpg)
+
+In order to build a valid custom `.cc.json`, these rules must be followed:
```
```suggestion
The following rules apply in order to build a valid custom `.cc.json`:
```

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
```diff
@@ -0,0 +1,101 @@
+-   The value `left + width` or `top + height` must be in a range of `[0, 100]`.
+-   Folders may not overlap.
+-   Leaving space between folders (for visibility reasons) is recommended.
+-   Only folders on the root can have the `fixedPosition`-attribute.
+-   If there exists at least on fixated folder, all folders on the root must be fixated.
```
```suggestion
-   All folders on the root level have to use the `fixedPosition` attribute.
```

NearW replied:
`on the root` still sounds misleading to me.
What about:
```suggestion
-   All direct sub-folders of the root have to use the `fixedPosition` attribute.
```

BridgeAR replied:
```suggestion
-   All root folders have to use the `fixedPosition` attribute.
```

What about keeping it brief?

NearW replied:
There is only one root folder and it should not have the `fixedPosition` attribute. All direct children of that folder require it though, if at least one of them has it.

BridgeAR replied:
```suggestion
-   All children of the root folder require the `fixedPosition` attribute.
```

How about that? (I explicitly did not mention "if at least one of them has the attribute set". This whole section is only about the fixed position)

BridgeAR commented on gh-pages/_posts/how-to/2020-08-17-fixate-folders-with-a-custom-cc-json.md:
````diff
@@ -0,0 +1,101 @@
+}
+```
+
+Since a margin between folders is recommended, we set the coordinates for the folder in the top-left corner
+to `left: 5` and `top: 5`. The margin between the border of the map and `folder_1` is therefore `5`, but can be chosen at will.
+In order to set the coordinates to adjacent folders, we can again apply any margin between `folder_1` and `folder_2`. We chose
+a margin of `5` again, so that the coordinates of `folder_2` must be `left: 60` and `top: 40`.
````
```suggestion
A margin between folders is recommended. To apply one, set the coordinates for the folder in the top-left corner
to e.g., `left: 5` and `top: 5`. The margin between the border of the map and `folder_1` is therefore `5`. It may be chosen at will.
In order to define the coordinates of adjacent folders, apply any margin between `folder_1` and `folder_2`. The example uses a margin of `5`, so that the coordinates of `folder_2` must be `left: 60` and `top: 40`.
```

BridgeAR reviewed (approved):
LGTM, just left a few suggestions.

NearW commented:
LGTM now, too. :b

NearW commented:
Updated the example image with something more descriptive. I also updated the `cc.json` example to match this image.

![fixated-folder-example](https://user-images.githubusercontent.com/12533626/92637800-a77c7d00-f2d9-11ea-99a6-f36e83f009b2.jpg)
